sbi/inference/nspe/nspe.py

```diff
@@ -0,0 +1,633 @@
+# This file is part of sbi, a toolkit for simulation-based inference. sbi is licensed
+# under the Affero General Public License v3, see <https://www.gnu.org/licenses/>.
+import time
+from copy import deepcopy
+from typing import Any, Callable, Optional, Union
+from warnings import warn
+
+import torch
+from torch import Tensor, ones, optim
+from torch.distributions import Distribution
+from torch.nn.utils.clip_grad import clip_grad_norm_
+from torch.utils.tensorboard.writer import SummaryWriter
+
+from sbi import utils as utils
+from sbi.inference import NeuralInference, check_if_proposal_has_default_x
+from sbi.inference.posteriors import (
+    DirectPosterior,
+)
+from sbi.inference.posteriors.base_posterior import NeuralPotentialPosterior
+from sbi.inference.posteriors.score_posterior import ScorePosterior
+from sbi.neural_nets.estimators.score_estimator import ConditionalScoreEstimator
+from sbi.neural_nets.factory import posterior_score_nn
+from sbi.utils import (
+    RestrictedPrior,
+    check_estimator_arg,
+    handle_invalid_x,
+    npe_msg_on_invalid_x,
+    test_posterior_net_for_multi_d_x,
+    validate_theta_and_x,
+    warn_if_zscoring_changes_data,
+    x_shape_from_simulation,
+)
+from sbi.utils.sbiutils import ImproperEmpirical, mask_sims_from_prior
+
+
+class NSPE(NeuralInference):
```

Comment:
The name is reasonable, but NSPE<-> SNPE might be confusing?

Reply:
yes we will rename SNPE to NPE (but in a separate PR)